keanu-python/keanu/algorithm/sampling.py

```diff
@@ -25,40 +26,65 @@ def sample(net: BayesNet,
            algo: str = 'metropolis',
            draws: int = 500,
            drop: int = 0,
-           down_sample_interval: int = 1) -> sample_types:
+           down_sample_interval: int = 1,
+           plot: bool = False,
+           ax: Any = None) -> sample_types:
 
     vertices_unwrapped = k.to_java_object_list(sample_from)
 
     network_samples = algorithms[algo].withDefaultConfig().getPosteriorSamples(
         net.unwrap(), vertices_unwrapped, draws).drop(drop).downSample(down_sample_interval)
     vertex_samples = {
-        Vertex._get_python_id(vertex_unwrapped): list(
+        Vertex._get_python_label(vertex_unwrapped): list(
             map(Tensor._to_ndarray,
                 network_samples.get(vertex_unwrapped).asList())) for vertex_unwrapped in vertices_unwrapped
     }
 
+    if plot:
+        traceplot(vertex_samples, ax=ax)
+
     return vertex_samples
 
 
 def generate_samples(net: BayesNet,
                      sample_from: Iterable[Vertex],
                      algo: str = 'metropolis',
                      drop: int = 0,
-                     down_sample_interval: int = 1) -> sample_generator_types:
+                     down_sample_interval: int = 1,
+                     live_plot: bool = False,
+                     refresh_every: int = 100,
+                     ax: Any = None) -> sample_generator_types:
     vertices_unwrapped = k.to_java_object_list(sample_from)
 
     sample_iterator = algorithms[algo].withDefaultConfig().generatePosteriorSamples(net.unwrap(), vertices_unwrapped)
     sample_iterator = sample_iterator.dropCount(drop).downSampleInterval(down_sample_interval)
     sample_iterator = sample_iterator.stream().iterator()
 
-    return _samples_generator(sample_iterator, vertices_unwrapped)
+    return _samples_generator(
+        sample_iterator, vertices_unwrapped, live_plot=live_plot, refresh_every=refresh_every, ax=ax)
+
 
+def _samples_generator(sample_iterator: JavaObject, vertices_unwrapped: JavaList, live_plot: bool, refresh_every: int,
+                       ax: Any) -> sample_generator_types:
+    traces = []
+    x0 = 0
 
-def _samples_generator(sample_iterator: JavaObject, vertices_unwrapped: JavaList) -> sample_generator_types:
     while (True):
         network_sample = sample_iterator.next()
         sample = {
-            Vertex._get_python_id(vertex_unwrapped): Tensor._to_ndarray(network_sample.get(vertex_unwrapped))
+            Vertex._get_python_label(vertex_unwrapped): Tensor._to_ndarray(network_sample.get(vertex_unwrapped))
             for vertex_unwrapped in vertices_unwrapped
         }
+
+        if live_plot:
+            traces.append(sample)
+            if len(traces) % refresh_every == 0:
+                joined_trace = {k: [t[k] for t in traces] for k in sample.keys()}
+                if ax is None:
```

Comment:
Why do we need the `if` here? We don't do it when we call `traceplot` from `sample`, we just pass the `None` `ax` straight to `traceplot`

Reply:
It's so that between sample batches, we draw to the same window (by using the same `ax`)